File: C++/110. Balanced Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int dfs(TreeNode* node)
    {
        if(node == nullptr)
            return 0;
        return 1 + max(dfs(node->left), dfs(node->right));
    }
    bool isBalanced(TreeNode* root) {
        if(root == nullptr)
            return true;
        int l = dfs(root->left);
        int r = dfs(root->right);
        if(abs(l-r) > 1)
            return false;
        return isBalanced(root->left) && isBalanced(root->right);
    }
};

File: Python/110. Balanced Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def dfs(self, node):
        if node is None:
            return 0
        return 1 + max(self.dfs(node.left), self.dfs(node.right))
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        if root is None:
            return True
        l = self.dfs(root.left)
        r = self.dfs(root.right)
        if abs(l - r) > 1:
            return False
        return self.isBalanced(root.left) and self.isBalanced(root.right)